[Ready for Review] Cas9-based Lineage Tracing Experiment Simulator

mattjones315 opened the pull request:
This PR implements the Cas9 based lineage tracing data simulator. The purpose of this simulator is to provide a realistic framework for generating data akin to the technologies described in Chan et al, Nature 2019 and McKenna et al, Science 2016. This includes, but is not limited to:

* segmentation of characters into equally sized "cassettes"
* integration of specific mutation rates, to simulate the heavy-tailed Cas9-indel distributions observed in real studies
* inclusion of multiple missing data sources -- specifically, heritable missing data and stochastic missing data
* resection events that are caused by simultaneous Cas9 cutting on the same cassette, resulting in missing data

The character states are populated in a `CassiopeaiTree` that is passed in with branch lengths.

All tests are stored in `tests/simulator_tests/cas9_lineage_tracing_simulator_tests.py`.

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,202 @@
+        Args:
+            character_array: Character array
+
+        Returns:
+            An updated character array.
```
Actually this returns nothing: the `character_array` is modified in-place.

mattjones315 replied:
Good catch! I changed this, and addressed the comment below by creating a copy that we modify and returning this modified copy.

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,202 @@
+            character_array: Character array
+            cuts: Loci being cut
+
+        Returns:
+            An updated character array.
```
This method also seems to modify the `character_array` in-place; we should document this.

mattjones315 replied:
See response above.

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,202 @@
+        """
+
+        for i in cuts:
+            state = self.random.choice(list(self.mutation_priors.keys()), 1, p=list(self.mutation_priors.values()))
```
This line is hard to understand, can we split it into multiple lines, each of which is easy to understand?

mattjones315 replied:
I fixed this in a recent commit, how does it look now?

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,202 @@
+        mutation_rate: float = 0.01,
+        state_distribution: Callable[None, float] = lambda: np.random.exponential(1e-5),
+        number_of_states: int = 100,
+        silencing_rate: float = 1e4,
```
1e-4?

mattjones315 replied:
Good catch!

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,202 @@
+            number_of_mutations = np.random.binomial(len(open_sites), p)
+            new_cuts = np.random.choice(open_sites, number_of_mutations)
+            
+            # collapse cuts that are on the same cassette
+            character_array, cuts_remaining = self.collapse_sites(character_array, new_cuts)
```
Hmm, this is not how I was expecting a double cut to happen. In my mind, whether a double cut happens or not should not depend on what cuts got _accumulated_ during the lifetime of a cell: a double cut should depend on whether the two cuts happened _close_ together in time.

mattjones315 replied:
I totally agree - this is not necessarily _exactly_ how double cuts would happen.

However, there are lots of other things that we are not modeling. We're not actually modeling DNA repair, we're not actually modeling how long Cas9 takes to induce a cut, etc. These all undoubtedly play a part in these multiple-site resections. I think a good approximation is to assume that if Cas9 cuts at two sites on the same recorder in a cell's lifetime, we'll introduce a resection event.

sprillo replied:
Sounds good! So long as we document this precisely.

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,202 @@
+            character_array = tree.get_character_states(parent)
+            open_sites = [c for c in range(len(character_array)) if character_array[c] == 0]
+
+            number_of_mutations = np.random.binomial(len(open_sites), p)
+            new_cuts = np.random.choice(open_sites, number_of_mutations)
```
Just to make sure: this is the same as tossing a coin with probability `p` for each open site?

richardyz98 replied:
yes!

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,202 @@
+    Args:
+        number_of_cassettes: Number of cassettes (i.e., arrays of target sites)
+        size_of_cassette: Number of editable target sites per cassette
+        mutation_rate: Exponential parameter for the Cas9 cutting rate.
```
Have you considered having a different mutation rate for each site?

mattjones315 replied:
We have done this experiment before -- you can see it in the "Phased Editor" part of our Genome Biology paper. We could easily add this to the simulator.

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,202 @@
+
+    Finally, the class accepts a silencing rate. This is normally a rare event
+    in which an entire cassette is transcriptionally silenced and therefore
+    not observed.
```
Maybe we should be more explicit that silencing happens when a cell divides? (As opposed to stochastic dropout)

mattjones315 replied:
Good call -- I'll add this to the documentation.

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,202 @@
+        number_of_cassettes: int = 10,
+        size_of_cassette: int = 3,
+        mutation_rate: float = 0.01,
+        state_distribution: Callable[None, float] = lambda: np.random.exponential(1e-5),
```
Why a callable and not just a list of floats?

mattjones315 replied:
Ah, that'd be nice. I'll add an option for this!

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,202 @@
+        size_of_cassette: Number of editable target sites per cassette
+        mutation_rate: Exponential parameter for the Cas9 cutting rate.
+        state_distribution: Distribution from which to simulate state
+            likelihoods
```
So, this is not something that sums to 1? We pull out a bunch of values and then normalize?

mattjones315 replied:
Yes, this is exactly how I am using it

mattjones315 replied:
This is a good point Richard, I've changed the variable name in my latest set of commits.

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,202 @@
+                for site in range(left, right+1):
+                    character_array[site] = -1
+
+    def get_cassettes(self) -> List[int]:
```
Should this (and other methods) be made private?

mattjones315 replied:
I prefer to keep these public for testing purposes. Is there a strong reason for why we'd not like a user to have access to these methods?

sprillo replied:
No strong reason, other than when reading code I like being able to distinguish key methods (public) from implementation details (private). But I can also infer this by looking at the abstract base class we are implementing, so not really an issue. In this case, you are right that there might be a reason for a user to want to use this method.

mattjones315 replied:
Okay, sounds good! Let's keep them public for now.

sprillo reviewed (commented):
I know this is a WIP, but since it was short I took a look out of curiosity and added some comments!

richardyz98 commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,327 @@
+                character_array, cuts_remaining
+            )
+
+            # silence cassettes
+            character_array = self.silence_cassettes(
+                character_array, self.heritable_silencing_rate
+            )
```
Is it true that silencing only happens at division events?

mattjones315 replied:
Yes and no. It might as well just happen at division events in the simulation -- because if the silencing happens at all during a cell's lifetime it's going to get passed on to its children. But more generally, we don't really know when the heritable silencing happens.

richardyz98 replied:
I guess what changes is our interpretation of the `heritable_silencing_rate` parameter. The way it's implemented now, it's dependent on the underlying tree topology (how many generations occur determines the probability that a silencing event will happen in any given character in any given cell). Whereas if we used the silencing rate as a true rate in an exponential process like we're doing for the mutation rate, it might be easier to build estimators for. (If a silencing event happens at all on a branch, then it'll just override any other mutation that might have occurred at a site on that branch in the case where we use an exponential).

sprillo replied:
Nice catch @richardyz98 . I agree with @richardyz98 , the `heritable_silencing_rate` is per division event and not per unit time. @mattjones315 which is more reasonable biologically? We should implement the more biologically reasonable one. I feel like per time unit (as @richardyz98 mentioned) is more reasonable: the longer the cell lives, the more likely a heritable dropout is.

mattjones315 replied:
This is related to a point I made above about the intersection between biology and the simulation -- I'm actually not entirely sure when the transcriptional silencing would happen (i.e., if it's at a division event or if it could occur during a cell's lifetime). I agree though, for the purposes of building estimators, let's encode this as a time-dependent process. I'm happy to add this to the simulator!

sprillo replied:
Gotcha, then I'm fine either way! Let's just remember to document it.

mattjones315 replied:
In my latest commit I've updated the tree simulator to have a time-dependence to the transcriptional silencing. here, I think it makes sense to model -- especially because we'll want to use some estimators here

mattjones315 commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,327 @@
+                character_matrix[node] = [0] * number_of_characters
+                continue
+
+            parent = tree.parent(node)
```
Sorry, I don't see how this can happen -- I'm checking specifically above whether or not the node is the root. You should never hit `parent = tree.parent(node)` if the node is the root.

richardyz98 replied:
I missed this! Thank you for clarifying

mattjones315 commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,327 @@
+                    new_cuts.append(site)
+
+            # collapse cuts that are on the same cassette
+            cuts_remaining = new_cuts
+            if self.size_of_cassette > 1:
+                character_array, cuts_remaining = self.collapse_sites(
+                    character_array, new_cuts
+                )
```
I was thinking about this too, actually. But if a user does not want this behavior, they should just pass `size_of_cassette = 1`

richardyz98 replied:
Ah that is true! Perhaps we can mention that in the docstring then

sprillo replied:
What if a user want to have different cut rates for each site of the cassette (a feature I requested in a comment above), and doesn't want collapsing of cuts? In this case, I like @richardyz98 's proposal to add a parameter that allows the user to specify avoiding this behavior.

sprillo replied:
Also, I mentioned this in my previous review, but I don't think this model of collapsing is biologically accurate. I am fine leaving it as it is, as long as we explain exactly how collapsing is handled: regardless of when then cuts happen during the lifetime of the cell (i.e. even if two cuts happen a very long time apart), collapsing will still happen when the cell divides.

mattjones315 replied:
There are two points here --

(1) **Cutting rates & an options for collapsing**. This code supports both of these with a lot of flexibility. A user can pass in an arbitrary list of mutation rates for their lineage tracer by passing in a list of floats to `mutation_rate`. This allows a user to specify the mutation rates with exquisite control -- per target site, per cassette, or a combination of both. Re: collapsing -- a user can encode this, *without changing any other behavior*, by specifying the `size_of_cassette` to be 1. You'll still get transcriptional silencing here but no resections.

(2) Your point is well-taken Sebastian. My point is that the biology behind double-resections is not well known at all, especially with respect to how long the Cas9 is residing while cutting and what the time window is. We cannot feasibly model everything that happens biologically; e.g.the random mutations that are introduced upon cell division, the opening and closing of DNA regions, the silencing of *guides*, etc. I think that this is a fair generalization to how resections are taking place. This is a general comment on the relationship between simulation & reality that I will also make below when we talk about the transcriptional silencing of the cassettes. In any event, I will make an addendum to the documentation to make this more clear!

sprillo replied:
Oh, I see, I didn't notice the user could specify a list of floats for the `mutations_rate`! We should update the docstring to clarify this, currently it says: `mutation_rate: Exponential parameter for the Cas9 cutting rate.`

I'm cool with the double-resections logic then!

mattjones315 replied:
great!

mattjones315 commented:
Hey @richardyz98, I believe I hit most of your comments (either by responding to them or by changing stuff in the code). Thanks for the feedback!

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,328 @@
+        if self.mutation_priors is None:
+            self.mutation_priors = {}
+            probabilites = [
+                state_generating_distribution() for _ in range(number_of_states)
```
This step happens before a random seed is set, so it might produce different results on different runs of the global script? I would probably move this code out of the `__init__` method and into the `overlay_data` method.

mattjones315 replied:
Ah, this is a great point! I'll move this to the `overlay_data` method.

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,328 @@
+            simulating heritable missing data events.
+        stochastic_silencing_rate: Rate at which to randomly drop out cassettes,
+            to simulate dropout due to low sensitivity of assays.
+        random_seed: Random seed to use for deterministic simulations. Note
```
I think we should be more precise and say that this sets the *numpy* random seed.

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,328 @@
+            new_cuts = []
+            for site in open_sites:
+                mutation_rate = self.mutation_rate_per_character[site]
+                p = 1 - (np.exp(-t * mutation_rate))
```
Can we call this variable `p` something more expressive, like `mutation_probability`?

mattjones315 replied:
Good point!

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,328 @@
+        for leaf in tree.leaves:
+            character_matrix[leaf] = self.silence_cassettes(
+                character_matrix[leaf], self.stochastic_silencing_rate
+            )
```
I think it would be *great* if we allowed the user to specify what integers to use to represent stochastic and heritable dropouts. That way, we can easily simulate data where we know whether a dropout in the character matrix is heritable or stochastic. This is really easy to implement: we just have to add two more arguments to `__init__`. I think as long as we require that they are negative numbers, and by default they are `-1`, we should be good. I know that some downstream methods assume that the missing state is `-1`, so we should definitely add a warning.

mattjones315 replied:
That's a good idea! I'll add this to the simulator.

sprillo reviewed (changes requested):
Added some comments (mostly feature requests). Let me know what you think!

mattjones315 commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,328 @@
+class Cas9LineageTracingDataSimulator(LineageTracingDataSimulator):
+    """Simulates Cas9-based lineage tracing data.
+
+    This subclass of `LineageTracingDataSimulator` extends the `overlay_data`
```
Good call -- I'll change this in the documentation!

mattjones315 commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,328 @@
+    
+    Second, the class accepts architecture of the recorder - described by the
+    size of the cassette (by default 3) and the number of cassettes. The
+    resulting lineage will have |cassette|*(# of cassettes) characters.
```
Good call

mattjones315 commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,328 @@
+            raise DataSimulatorError("Specify a positive number of cassettes.")
+        if size_of_cassette <= 0 or not isinstance(size_of_cassette, int):
+            raise DataSimulatorError(
+                "Specify a positive number of cut-sites" " per cassette."
```
To keep line lengths at 80 characters.

sprillo replied:
Isn't:
`"Specify a positive number of cut-sites" " per cassette."` equivalent to:
`"Specify a positive number of cut-sites per cassette."` ?
I'm still confused about this.

mattjones315 replied:
You just can't see it here in the Github rendering, but in the real code there's a line separating these two strings. So, the code is formatted to respect the 80 character limit.

mattjones315 commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,328 @@
+        if isinstance(mutation_rate, float):
+            if mutation_rate < 0:
+                raise DataSimulatorError(
+                    "Mutation rate needs to be" " non-negative."
```
Same as above, re: line length.

mattjones315 commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,328 @@
+        self,
+        number_of_cassettes: int = 10,
+        size_of_cassette: int = 3,
+        mutation_rate: Union[float, List[float]] = 0.01,
```
You probably just missed this, but I do support this by allowing the user to pass in a List of floats. Each entry corresponds to a mutation rate for a given character. This gives the user ultimate control over the mutation rates, and can be organized according to cassette, target-site per cassette, or something else more clever.

sprillo replied:
Yeah, I totally missed this Matt! As I mentioned above, let's make this clear in the docstring for the `mutation_rate` parameter.

mattjones315 commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,328 @@
+
+            if np.any(np.array(mutation_rate) < 0):
+                raise DataSimulatorError(
+                    "Mutation rate needs to be" " non-negative."
```
Same as above, re: line length.

richardyz98 reviewed (commented):
Hi Matt,
I think it's looking great now!

Did you see my comment in the slack by chance? Perhaps in a future iteration we can have it so that we can allow the simulator to track the ground truth mutations so that we can see what dropped out mutations should have corresponded to? This would be really helpful in the missing data work

mattjones315 commented:
Hi @sprillo and @richardyz98 --

I believe I've addressed all of your comments. These were all really helpful, so thank you!

Richard: regarding the missing data, I think this might introduce too much engineering into the simulation and also has consequences with respect to the lineage tracing simulator. I could see how we would do it, but I propose that we have a separate simulator that will "corrupt" the data for your purposes. In particular, you would simulate the data onto the tree with no collapsing and no missing data and then have a separate simulation process that would update the character matrix by introducing missing data. Would this work for your purpose?

If you think it would be a lot cleaner to have this baked into the current simulator, I think I know how to introduce this cleanly.

richardyz98 commented:
Hi Matt,
I think I know what you mean by adding a new simulator that "corrupts" existing trees with missing data. I think that might be a useful functionality to have, so I'll approve this and implement that later for my own purposes. Thanks for all of the great work here!

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,367 @@
+        number_of_cassettes: Number of cassettes (i.e., arrays of target sites)
+        size_of_cassette: Number of editable target sites per cassette
+        mutation_rate: Exponential parameter for the Cas9 cutting rate. The user
+            can optionally pass in a float, in which every site will mutate at
```
I don't think this is `optionally`, I would say `either` instead?

mattjones315 replied:
This is what I meant in this context; happy to change it to "either" for clarity

sprillo replied:
Oh, I see! I was thinking in the context of python types: `mutation_rate` is a `Union` rather than an `Optional`. I'm glad to leave as it is though, I'm just overloaded with mypy!

mattjones315 replied:
Haha, no worries! I changed it to "either" for clarity.

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,367 @@
+    ) -> List[int]:
+        """Silences cassettes.
+
+        Using the specied silencing rate, this function will randomly select
```
typo: `specied`

sprillo commented on cassiopeia/simulator/Cas9LineageTracingDataSimulator.py:
```diff
@@ -0,0 +1,367 @@
+                raise DataSimulatorError(
+                    "Mutation rate needs to be" " non-negative."
+                )
+            self.mutation_rate_per_character = mutation_rate
```
Let's make a copy to avoid exposing the internal representation of the class (similar to the issue we had with CassiopeiaTree). E.g.: `self.mutation_rate_per_character = mutation_rate[:]`

mattjones315 replied:
Good call, this is done in my latest commit.
